test/Test/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -54,9 +63,175 @@ tests = testGroup "Database.LSMTree.Internal.RawPage"
+        rawPageKeys page @=? V.fromList [P.fromList [0x42, 0x43], P.fromList [0x52, 0x53]]
+        rawPageValues page @=? V.fromList [P.fromList [0x44, 0x45], P.fromList [0x54, 0x55]]
+
+        rawPageEntry page (P.fromList [0x52, 0x53]) @=? Just (Entry.Mupdate (P.fromList [0x54,0x55]))
+        rawPageEntry page (P.fromList [0x99, 0x99]) @=? Nothing
+
+    , testProperty "keys" prop_keys
+    , testProperty "values" prop_values
+    , testProperty "hasblobrefs" prop_hasblobrefs
+    , testProperty "blobrefs" prop_blobrefs
+    , testProperty "ops" prop_ops
+    , testProperty "entries" prop_entries_exists
     ]
 
 toRawPage :: PageLogical -> RawPage
 toRawPage p = case SBS.toShort $ serialisePage $ encodePage p of
     SBS.SBS ba -> makeRawPage (ByteArray ba) 0
+
+prop_keys :: PageLogical -> Property
+prop_keys p@(PageLogical xs) =
+    rawPageKeys rawpage === V.fromList keys
+  where
+    rawpage = toRawPage p
+
+    keys :: [P.Vector Word8]
+    keys = [ P.fromList (BS.unpack bs) | (Key bs, _, _) <- xs ]
+
+prop_values :: PageLogical -> Property
+prop_values p@(PageLogical xs) =
+    rawPageValues rawpage === V.fromList values
+  where
+    rawpage = toRawPage p
+
+    values :: [P.Vector Word8]
+    values = [ extractValue op | (_, op, _) <- xs ]
+
+    extractValue (Insert (Value bs))  = bsToVector bs
+    extractValue (Mupsert (Value bs)) = bsToVector bs
+    extractValue Delete               = P.empty
+
+prop_blobrefs :: PageLogical -> Property
+prop_blobrefs p@(PageLogical xs) =
+    [rawPageBlobRefIndex rawpage i | i <- [0 .. length blobRefs - 1] ] === blobRefs
+  where
+    rawpage = toRawPage p
+
+    blobRefs :: [(Word64,Word32)]
+    blobRefs = [ (x, y) | (_, _, Just (BlobRef x y)) <- xs ]
+
+prop_hasblobrefs :: PageLogical -> Property
+prop_hasblobrefs p@(PageLogical xs) =
+    [ rawPageHasBlobRefAt rawpage i /= 0 | i <- [0 .. length xs - 1] ] === blobRefs
+  where
+    rawpage = toRawPage p
+
+    blobRefs :: [Bool]
+    blobRefs = [ isJust mb | (_, _, mb) <- xs ]
+
+prop_ops :: PageLogical -> Property
+prop_ops p@(PageLogical xs) =
+    [ rawPageOpAt rawpage i | i <- [0 .. length xs - 1] ] === ops
+  where
+    rawpage = toRawPage p
+
+    ops :: [Word64]
+    ops = [ fromOp o | (_, o, _) <- xs ]
+
+    fromOp :: Operation -> Word64
+    fromOp Insert {}  = 0
+    fromOp Delete {}  = 2
+    fromOp Mupsert {} = 1
+
+prop_entries_exists :: PageLogical -> Property
+prop_entries_exists   (PageLogical [])  = property True
+prop_entries_exists   (PageLogical xs') = forAll (elements xs) $ \(Key k, op, blobref) ->
+    rawPageEntry rawpage (bsToVector k) === Just case op of
+        Insert (Value v)       -> case blobref of
+            Nothing            -> Entry.Insert (bsToVector v)
+            Just (BlobRef x y) -> Entry.InsertWithBlob (bsToVector v) (x, y)
+        Mupsert (Value v)      -> Entry.Mupdate (bsToVector v)
+        Delete                 -> Entry.Delete
+  where
+    rawpage = toRawPage (PageLogical xs)
```

Comment:
Maybe add a negative test too -- if the key isn't in the page, then it should return Nothing